api/desecapi/models.py

```diff
@@ -156,56 +134,91 @@ def pdns_id(self):
         """
         Make sure that pdns gets the latest information about this domain/zone.
         Re-Syncing is relatively expensive and should not happen routinely.
         """
 
         # Create zone if it does not exist yet
         try:
+            # This will also import any RRsets that may have been created already
             pdns.create_zone(self)
         except pdns.PdnsException as e:
-            if not (e.status_code == 422 and e.detail.endswith(' already exists')):
-                raise e
-
-        # update zone to latest information
-        pdns.set_dyn_records(self)
-
-    def pdns_sync(self, new_domain):
-        """
-        Command pdns updates as indicated by the local changes.
-        """
+            if (e.status_code == 422 and e.detail.endswith(' already exists')):
+                # Delete all RRsets from pdns
+                pdns.set_rrsets(self, [])
 
-        if self.owner.captcha_required:
-            # suspend all updates
-            return
-
-        # if this zone is new, create it and set dirty flag if necessary
-        if new_domain:
-            pdns.create_zone(self)
-            self._dirtyRecords = bool(self.arecord) or bool(self.aaaarecord) or bool(self.acme_challenge)
-
-        # make changes if necessary
-        if self._dirtyRecords:
-            pdns.set_dyn_records(self)
-
-        self._dirtyRecords = False
+                # Update zone to latest information
+                pdns.set_rrsets(self, self.get_rrsets())
+            else:
+                raise e
 
     @transaction.atomic
     def sync_from_pdns(self):
         RRset.objects.filter(domain=self).delete()
         for rrset in pdns.get_rrsets(self):
             if rrset['type'] not in RRset.RESTRICTED_TYPES:
                 RRset(**rrset).save(pdns=False)
+        ## TODO is this better?
+        # rrsets = pdns.get_rrsets(self)
+        # rrsets = [rrset for rrset in rrsets if
+        #           rrset.type not in RRset.RESTRICTED_TYPES]
+        # RRset.objects.bulk_create(rrsets)
+
+    @transaction.atomic
+    def set_rrsets(self, rrsets):
+        for rrset in rrsets:
+            if rrset.domain != self:
+                raise ValueError(
+                    'Cannot set RRset for domain %s on domain %s.' % (
+                    rrset.domain.name, self.name))
+            if rrset.type in RRset.RESTRICTED_TYPES:
+                raise ValueError(
+                    'You cannot tinker with the %s RRset.' % rrset.type)
+
+        for rrset in rrsets:
+            # Look up old RRset to see if it needs updating.  If exists and
+            # outdated, delete it so that we can bulk-create it later.
+            qs = RRset.objects.filter(domain=self.pk, subname=rrset.subname,
+                                      type=rrset.type)
+
+            # Identical RRset exists; no need to recreate
+            if RR.objects.filter(rrset__in=qs, rrset__ttl=rrset.ttl).values('content') == rrset.records_data:
```

Comment:
may have false positives due to ordering

Reply:
fixed